imagemanager/imagemanager.go

```diff
@@ -0,0 +1,1076 @@
+	if err = imagemanager.removeServicesIfOutdated(); err != nil {
+		log.Errorf("Can't remove outdated services: %v", err)
+	}
+
+	if err = imagemanager.removeLayersIfOutdated(); err != nil {
+		log.Errorf("Can't remove outdated layers: %v", err)
+	}
+
+	go imagemanager.validateTTLs()
+
+	return imagemanager, nil
+}
+
+// Close close image manager object.
+func (imagemanager *Imagemanager) Close() {
+	if err := imagemanager.serviceAllocator.Close(); err == nil {
+		log.Errorf("Can't close service allocator: %v", err)
+	}
+
+	if err := imagemanager.layerAllocator.Close(); err == nil {
+		log.Errorf("Can't close layer allocator: %v", err)
+	}
+
+	if err := imagemanager.tmpAllocator.Close(); err == nil {
+		log.Errorf("Can't close tmp allocator: %v", err)
+	}
+
+	close(imagemanager.validateTTLStopChannel)
+}
+
+// GetServicesStatus gets all services status.
+func (imagemanager *Imagemanager) GetServicesStatus() ([]unitstatushandler.ServiceStatus, error) {
+	log.Debug("Get services status")
+
+	servicesInfo, err := imagemanager.storage.GetServicesInfo()
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	servicesStatus := make([]unitstatushandler.ServiceStatus, len(servicesInfo))
+
+	for i, service := range servicesInfo {
+		servicesStatus[i] = unitstatushandler.ServiceStatus{
+			ServiceStatus: cloudprotocol.ServiceStatus{
+				ID:         service.ID,
+				AosVersion: service.AosVersion,
+				Status:     cloudprotocol.InstalledStatus,
+			}, Cached: service.Cached,
+		}
+	}
+
+	return servicesStatus, nil
+}
+
+// GetLayersStatus gets all layers status.
+func (imagemanager *Imagemanager) GetLayersStatus() ([]unitstatushandler.LayerStatus, error) {
+	log.Debug("Get layers status")
+
+	layersInfo, err := imagemanager.storage.GetLayersInfo()
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	layersStatus := make([]unitstatushandler.LayerStatus, len(layersInfo))
+
+	for i, layer := range layersInfo {
+		layersStatus[i] = unitstatushandler.LayerStatus{
+			LayerStatus: cloudprotocol.LayerStatus{
+				ID:         layer.ID,
+				AosVersion: layer.AosVersion,
+				Digest:     layer.Digest,
+				Status:     cloudprotocol.InstalledStatus,
+			}, Cached: layer.Cached,
+		}
+	}
+
+	return layersStatus, nil
+}
+
+// InstallService install service to the image store dir.
+func (imagemanager *Imagemanager) InstallService(serviceInfo cloudprotocol.ServiceInfo,
+	chains []cloudprotocol.CertificateChain, certs []cloudprotocol.Certificate,
+) error {
+	log.WithFields(log.Fields{"id": serviceInfo.ID}).Debug("Install service")
+
+	serviceFromStorage, err := imagemanager.storage.GetServiceInfo(serviceInfo.ID)
+	if err != nil && !errors.Is(err, ErrNotExist) {
+		return aoserrors.Wrap(err)
+	}
+
+	var isServiceExist bool
+
+	if err == nil {
+		isServiceExist = true
+
+		if serviceInfo.AosVersion <= serviceFromStorage.AosVersion {
+			return ErrVersionMismatch
+		}
+	}
+
+	decryptedFile := path.Join(imagemanager.servicesDir, serviceInfo.ID+decryptedFileExt)
```

Comment:
Service ID is not suitable here, as we may have two service versions same time.

Reply:
Done